File: src/modules/authentication/services/auth_recover_password_service.py
import os
from src.modules.authentication.domain.dto.auth_recover_password_dto import (
    AuthRecoverPasswordDTO,
)
from src.modules.user.models import User
from src.shared.exceptions.not_found_exception import NotFoundException
from src.shared.services.token_sign_service import TokenSignService
from src.shared.domain.enums.period_expiration_enum import PeriodExpiration
from src.modules.mail.services.mail_service import MailService
from src.modules.mail.domain.enums.mail_template_enum import EmailTemplate
from src.modules.mail.domain.interfaces.mail_option_interface import IMailOption
from src.modules.authentication.domain.interfaces.auth_recover_password_interface import (
    IAuthRecoverPassword,
)


class AuthRecoverPasswordService:
    def execute(dto: AuthRecoverPasswordDTO):
        user = User.objects.filter(email=dto.email, deleted_at=None).first()

        if user is None:
            raise NotFoundException("E-mail incorreto!")

        token = TokenSignService.execute(
            user=user, time=10, period=PeriodExpiration.MINUTE
        )
        url = (
            f'{os.getenv("FRONT_END_URL", "")}/recuperar-senha/nova-senha?token={token}'
        )

        context: IAuthRecoverPassword = {"name": user.name, "url": url}
        options: IMailOption = {
            "to": user.email,
            "subject": "Recuperação de Senha:",
            "template": EmailTemplate.RECOVER_PASSWORD,
            "context": context,
        }
        MailService.send(options=options)

        return {"message": "E-mail enviado com sucesso!"}

File: src/modules/auth/services/auth-recover-password.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { PrismaService } from 'src/infra/database/prisma.service';
import { TokenSignService } from 'src/shared/services/token-sign.service';
import { AuthRecoverPasswordDto } from '../domain/dto/auth-recover-password.dto';
import { MailService } from 'src/modules/mail/services/mail.service';
import { IMailOption } from 'src/modules/mail/domain/interfaces/mail-option.interface';
import { EMailTemplate } from 'src/modules/mail/domain/enums/mail-template.enum';
import { AuthRecoverPasswordResponseDto } from '../domain/dto/auth-recover-password-response.dto';
import { IAuthRecoverPassword } from '../domain/interfaces/auth-recover-password.interface';

@Injectable()
export class AuthRecoverPasswordService {
  constructor(
    private prisma: PrismaService,
    private tokenSignService: TokenSignService,
    private mailService: MailService,
  ) {}
  async execute(
    dto: AuthRecoverPasswordDto,
  ): Promise<AuthRecoverPasswordResponseDto> {
    const user = await this.prisma.user.findFirst({
      where: {
        email: dto.email,
        deletedAt: null,
      },
    });

    if (!user) throw new NotFoundException('E-mail incorreto!');

    const token = await this.tokenSignService.execute(user, '10m');

    const url = `${process.env.FRONT_END_URL}/recuperar-senha/nova-senha?token=${token}`;

    const options: IMailOption<IAuthRecoverPassword> = {
      to: user.email,
      subject: 'Recuperação de Senha:',
      template: EMailTemplate.RECOVER_PASSWORD,
      context: {
        name: user.name,
        url: url,
      },
    };

    await this.mailService.send(options);

    return { message: 'E-mail enviado com sucesso!' };
  }
}
